### PR TITLE
Enable node.js stack locals capture

### PR DESCRIPTION
**Pull request by waltjones**
## Description of the change

Enables capture of local stack variables in Node.js. To enable, set `locals: true` in the config.

This feature uses the Node Inspector API to access debugger data at runtime. The Inspector API is a wrapper for the V8 Inspector interface. While the V8 interface relies on `send()` and `receive()` functions that can't be accessed from Node, the Node Inspector API provides access to these via the `post()` method on the Inspector Session instance. Using `session.post()`, all of the DevTools protocol can be accessed, including the `Debugger` and `Runtime` interfaces, which are used here.

The initial version of this feature supports both caught and uncaught exceptions and promise rejections. Nested exceptions and trace chains are supported.

The DevTools protocol allows retrieving nested variables to an arbitrary depth. However, the initial version of this feature only implements the equivalent of depth = 0. A future PR may enable arbitrary depth with a config option to set the depth.

Unexpanded objects are identified by their class, in keeping with the example in the Rollbar API docs. (https://explorer.docs.rollbar.com/#tag/Item)

Example:
```
trace: {
  frames: [
    {
      // ...
      locals: {
        foo: 'bar', // immediate string value
        obj: '<Foo object>' // instance of Foo
      }
    }
  ]
}
```

The feature is enabled for Node 10+. Node 8 doesn't implement the `url` property on the `CallFrame` object, and cannot reliably resolve to the correct locations in the stack. (There may be a workaround, but none that I know at this time.)

This PR adheres to the established style of Rollbar.js and uses `var` declarations, rather than `let` and `const`. ES5 style classes are used. I'm not sure whether `session.post()` supports async/await. Regardless, callbacks are used throughout as shared code that cannot be made async prevents the use of async/await in this code path.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

Fixes: ch74679

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [x]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [x] Changes have been reviewed by at least one other engineer
- [x] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:87]: same comment about this function as the above for `firstFrame` - this function isn't making use of `this` - does it need to be exported on the `Locals.prototype`?

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:108]: seems that you can pass `callback` as the thrid arg to `async.each` rather than having the wrapping function - is there a reason for the extra wrapping function?

seems it will be called no matter the outcome the way it's written

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:107]: does this function need to be attached to the `Locals.prototype` and exported? I don't see it used outside of the `mergeLocals` body which does need to be exported (public).

As for the `bind` call to this, that could be done by `mergeLocals` when passing maybe? Just a thought on trying to expose less of these internal methods if they're not needed externally

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:119]: why not filter before the loop using a `scopes.filter(scope => scope.type === 'local')`

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:149]: does this need to be exported as attached to the `Locals.prototype`?

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:137]: since the only ref to `this` is for the `this.getObjectValue` call, does this need to be exported on the `Locals.prototype`?

File: src/server/locals.js
Comment (eudaimos) [inline src/server/locals.js:5]: defining classes like this the old school JS way has the problem of possible function call without instantiation, i.e. missing the `new` keyword by just calling `Locals(…)` as a function. To prevent this from doing harm, a pattern emerged for checking how a constructor is called like this:

```javascript
function Locals(config) {
  if (!(this instanceof Locals)) {
    return new Locals(config);
  }
  // rest of constructor logic
}
```

**Review by eudaimos — COMMENTED** [review]
@waltjones 

Overall I made comments/asked questions about the implementation and public api of `Locals`

Also wondering, should there be unit tests for the `Locals` API?
The tests I see seem to be routed through the rollbar SDK

**Review by eudaimos — CHANGES_REQUESTED** [review]
I never seem to get this right - requested changes are comments in previous "Comment" and submitting this now as a Review for the process to work

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js]: Updated to use `find` and expect one local scope.

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js]: Constant and explanatory comment added.

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js]: Good catch. Each of these cases have been condensed now.

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js]: Non-interface functions are all private now.

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js:5]: Added.

**Comment by waltjones** [comment]
Major updates are:
* The inspector session is now a singleton at `Locals.session`. This makes more sense, since multiple Rollbar.js instances should not open multiple sessions. This has the side effect of greatly reducing the need to pass `this` around since the main need for it was when calling `session.post()`.
* Unit tests have been added, which allows exercising some conditions not produced when the actual inspector API is called.

**Comment by christopherseaman** [comment]
Update?

@eudaimos @waltjones

**Comment by waltjones** [comment]
Just waiting on approval or further feedback.

**Review by bishopb — APPROVED** [review]

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js:30]: For now, the codebase is standardized on `var` rather than `let`/`const`, though I think there are a few `const` instances that have been merged in the past. My preference is to stay consistent until the decision is made to move away from `var`.

The value itself shouldn't need to be customized. That's the intent anyway. If there were a demonstrated need (not just for it to be larger, but for it to be larger only in some cases), it could be parameterized then. It's much easier to add to the interface, than to remove or change something.

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js:47]: Updated to always return a map. I've kept the check at: https://github.com/rollbar/rollbar.js/blob/aaafe97927a0c0741300725330269f98dd84985c/src/server/parser.js#L313 

because `item.localsMap` will be undefined when locals is disabled.

File: src/server/locals.js
Comment (waltjones) [reply src/server/locals.js:49]: Addressed above.

**Pull request merged**
